### DISCUSSION TITLE
Unable to define query param of type list using pydantic model

### DISCUSSION THREAD

**Discussion by Pacheco95** [fastapi/fastapi#10090] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# Working in v0.95.1 but not in v0.101.1

from typing import Annotated

from fastapi import FastAPI, Query, Depends
from pydantic import BaseModel

app = FastAPI()


class QueryParams(BaseModel):
    q: Annotated[list[str], Query()]


@app.get("/")
def handler(q: QueryParams = Depends()):
    return q
```

### Description

I have this code in the example section that works in `v0.95.1`. I tried to upgrade to `v0.101.1` and it's now raising an error saying the handler expects a request body.

```json
{
  "detail": [
    {
      "type": "missing",
      "loc": [
        "body"
      ],
      "msg": "Field required",
      "input": null,
      "url": "https://errors.pydantic.dev/2.1/v/missing"
    }
  ]
}
```
How can I define a query param of type `list[str]` inside a pydantic model in FastAPI `v0.101.1`?

### Operating System

Linux

### Operating System Details

```
➜ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.3 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.3 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```

### FastAPI Version

0.101.1

### Pydantic Version

2.1.1

### Python Version

3.11.4

### Additional Context

I googled for a sollution in newer pydantic/fastapi versions but found nothing.

Searching for open issues I found these:

https://github.com/tiangolo/fastapi/issues/4413
https://github.com/tiangolo/fastapi/issues/9710

But none of them seems my case

**Comment by JoshYuJump** [comment]
#4413 is same of your case, but been closed with converted to discussion. Maybe it's a bug. 

It's the issue of Pydantic V2.

I have test it in FastAPI 0.95 / 0.96 / 0.98 / 0.100. It works fine in Pydantic 1.10, but failed in Pydantic V2.

**Comment by JoshYuJump** [comment]
When I did a more in-depth research. I notice it maybe a bug of Pydantic v2.

FastAPI inspect `QueryParams` using 
```
signature = inspect.signature(call)
```

In Pydantic v1, signature value is 
```
<Signature (*, q: typing.Annotated[list[str], Query(PydanticUndefined)]) -> None>
```

While in Pydantic v2, signature value is 
```
<Signature (*, q: list[str]) -> None>
```

Pydantic v2 has some magic make Python's inspect didn't work proper.

**Comment by YuriiMotov** [comment, marked as answer]
Since FastAPI 0.115.0 it's officially supported to use Pydantic models to define `Query` parameters: https://fastapi.tiangolo.com/tutorial/query-param-models/#query-parameters-with-a-pydantic-model
